### ISSUE TITLE
Complexity Limits / DDOS Circuit Breaker

### ISSUE THREAD

**Issue by AlecRosenbaum** [jg-rp/liquid#58]
Hi!

First of all, awesome library. Kudos!

I was wondering if there's any good way to guard against malicious input like this:
```
{% for i in (1..50000000000) %}
g
{% endfor %}
```

While this won't leak data, if a user creates a template like this it could bind up web workers, background job queues, etc.

I assume it's hard to know about this type of thing until render-time because of how variables are resolved within loops, but is there a good spot to put some sort of circuit-breaker complexity counter to prevent effectively infinite looping?

Perhaps:
* store an optional `max_complexity` within the `Environment`
* add a check within [for loop rendering](https://github.com/jg-rp/liquid/blob/e260cfa17b4cd71b4f7654d398cf0c9b6481ec39/liquid/builtin/tags/for_tag.py#L181) to check cumulative length against max_complexity, like:
```python
cumulative_length = functools.reduce(lambda x, y: x * y, [fl.length for fl in context.loops], length)
if cumulative_length > context.env.max_complexity:
	raise ...
```

**Comment by jg-rp** [comment]
Hi @AlecRosenbaum,

Yes, I think in [`Context.loop`](https://github.com/jg-rp/liquid/blob/main/liquid/context.py#L607), after the new `ForLoop` has been pushed onto the loop stack.

A slight variation on your example:

```python
    from operator import mul

    # ...

    @contextmanager
    def loop(self, namespace: Namespace, forloop: ForLoop) -> Iterator[Context]:
        """Just like ``Context.extend``, but keeps track of ForLoop objects too."""
        self.loops.append(forloop)

        if (
            reduce(mul, (loop.length for loop in self.loops))
            > self.env.loop_iteration_limit
        ):
            raise LoopIterationLimitError("loop iteration limit reached")

        with self.extend(namespace) as context:
            try:
                yield context
            finally:
                self.loops.pop()
```

There are some corner cases to consider. The `include`, `render` and `tablerow` tags all have the potential to loop over things, possibly causing `loop_iteration_limit` to be exceeded. Nothing a little refactoring can't solve though.

I am mindful of the lack of "resource limits" (the term Ruby Liquid uses) in Python Liquid, and took steps towards similar features when writing [LiquidScript](https://github.com/jg-rp/liquidscript). Steps that I intend to implement in Python Liquid in the future. Along with an output stream limit (in bytes), a loop limit is defiantly a feature I would like to include in Python Liquid.

## Working Example

Ignoring `include`, `render` and `tablerow` loops for a minute, the following will work now, with Python Liquid version 1.3.0.

```python
from contextlib import contextmanager
from functools import reduce
from operator import mul

from typing import Iterator

from liquid import Environment
from liquid import Context

from liquid.context import Namespace
from liquid.builtin.tags.for_tag import ForLoop
from liquid.exceptions import Error
from liquid.template import BoundTemplate


class LoopIterationLimitError(Error):
    """Exception raised when the maximum number of loop iterations is reached."""


class ContextWithLimits(Context):
    loop_iteration_limit = 1000

    @contextmanager
    def loop(self, namespace: Namespace, forloop: ForLoop) -> Iterator[Context]:
        """Just like ``Context.extend``, but keeps track of ForLoop objects too."""
        self.loops.append(forloop)

        if (
            reduce(mul, (loop.length for loop in self.loops))
            > self.loop_iteration_limit
        ):
            raise LoopIterationLimitError("loop iteration limit reached")

        with self.extend(namespace) as context:
            try:
                yield context
            finally:
                self.loops.pop()


class BoundTemplateWithLimits(BoundTemplate):
    context_class = ContextWithLimits


# Example usage
env = Environment()
env.template_class = BoundTemplateWithLimits

t = env.from_string(
    """\
{% for x in (1..100) -%}
{% for y in (1..11) -%}
{{x}},{{y}}
{% endfor -%}
{% endfor -%}
"""
)

print(t.render())
# LoopIterationLimitError: loop iteration limit reached, on line 1
```

**Comment by AlecRosenbaum** [comment]
Thank you for such a prompt and complete response! That should completely satisfy our usecase. The working example is super helpful, and makes it easy to integrate this functionality into the rendering process.

If leaving this issue open is helpful to track "resource limits" as a feature then I'm happy to leave it, but otherwise your reply has solved my issue. 🙂

**Comment by jg-rp** [comment]
:thumbsup: Lets keep it open.

**Comment by jg-rp** [comment]
Released in Python Liquid version 1.4.0. ([docs](https://jg-rp.github.io/liquid/guides/resource-limits))
